sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala

```diff
@@ -0,0 +1,170 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.streaming
+
+import java.io.File
+
+import org.apache.spark.sql.{DataFrame, Dataset}
+import org.apache.spark.sql.catalyst.plans.logical.Range
+import org.apache.spark.sql.catalyst.util.stringToFile
+import org.apache.spark.sql.connector.read.streaming
+import org.apache.spark.sql.connector.read.streaming.{ReadLimit, SupportsAdmissionControl}
+import org.apache.spark.sql.execution.streaming.{LongOffset, MemoryStream, Offset, SerializedOffset, Source, StreamingExecutionRelation}
+import org.apache.spark.sql.types.{LongType, StructType}
+
+class TriggerAvailableNowSuite extends FileStreamSourceTest {
+
+  import testImplicits._
+
+  abstract class TestDataFrameProvider {
+    @volatile var currentOffset = 0L
+
+    def toDF: DataFrame
+
+    def incrementAvailableOffset(numNewRows: Int): Unit
+  }
+
+  class TestSource extends TestDataFrameProvider with Source {
+    override def getOffset: Option[Offset] = {
+      if (currentOffset <= 0) None else Some(LongOffset(currentOffset))
+    }
+
+    override def getBatch(start: Option[Offset], end: Offset): DataFrame = {
+      if (currentOffset == 0) currentOffset = getOffsetValue(end)
+      val plan = Range(
+        start.map(getOffsetValue).getOrElse(0L) + 1L, getOffsetValue(end) + 1L, 1, None,
+        isStreaming = true)
+      Dataset.ofRows(spark, plan)
+    }
+
+    override def incrementAvailableOffset(numNewRows: Int): Unit = {
+      currentOffset += numNewRows
+    }
+
+    override def toDF: DataFrame =
+      Dataset.ofRows(spark, StreamingExecutionRelation(this, spark))
+
+    override def schema: StructType = new StructType().add("value", LongType)
+
+    override def stop(): Unit = {}
+
+    private def getOffsetValue(offset: Offset): Long = {
+      offset match {
+        case s: SerializedOffset => LongOffset(s).offset
+        case l: LongOffset => l.offset
+        case _ => throw new IllegalArgumentException("incorrect offset type: " + offset)
+      }
+    }
+  }
+
+  class TestSourceWithAdmissionControl extends TestSource with SupportsAdmissionControl {
+    override def getDefaultReadLimit: ReadLimit = ReadLimit.maxRows(1)  // this will be overridden
+
+    override def latestOffset(startOffset: streaming.Offset, limit: ReadLimit): streaming.Offset = {
+      val currentOffset = getOffset
+      assert(currentOffset.nonEmpty,
+        "the latestOffset should be called after incrementAvailableOffset")
+      currentOffset.get
+    }
+  }
+
+  class TestMicroBatchStream extends TestDataFrameProvider {
+
+    private lazy val memoryStream = MemoryStream[Long](0, spark.sqlContext)
+
+    override def toDF: DataFrame = memoryStream.toDF()
+
+    override def incrementAvailableOffset(numNewRows: Int): Unit = {
+      for (_ <- 1 to numNewRows) {
+        currentOffset += 1
+        memoryStream.addData(currentOffset)
+      }
+    }
+  }
+
```

Comment:
Let's make the test suite be exhaustive; 1) single source which doesn't support Trigger.AvailableNow 2) single source which supports Trigger.AvailableNow.

Latter can be covered by new FileStreamSource test, hence adding test for 1) would be enough.

Reply:
Sure. Will do.